client/asset/dcr/dcr.go

```diff
@@ -603,36 +635,48 @@ func (dcr *ExchangeWallet) SignMessage(coin asset.Coin, msg dex.Bytes) (pubkeys,
 	if err != nil {
 		return nil, nil, fmt.Errorf("error converting coin: %v", err)
 	}
-	tx, err := dcr.node.GetTransaction(&output.txHash)
-	if err != nil {
-		return nil, nil, err
-	}
-	for _, txDetails := range tx.Details {
-		if txDetails.Vout == output.vout &&
-			(txDetails.Category == txCatReceive ||
-				txDetails.Category == txCatGenerate) {
 
-			addr, err := dcrutil.DecodeAddress(txDetails.Address, chainParams)
-			if err != nil {
-				return nil, nil, fmt.Errorf("error decoding address: %v", err)
-			}
-
-			priv, pub, err := dcr.getKeys(addr)
-			if err != nil {
-				return nil, nil, fmt.Errorf("error fetching keys: %v", err)
-			}
-
-			signature, err := priv.Sign(msg)
-			if err != nil {
-				return nil, nil, fmt.Errorf("signing error: %v", err)
+	// First check if we have the funding coin cached. If so, grab the address
+	// from there.
+	dcr.fundingMtx.RLock()
+	fCoin, found := dcr.fundingCoins[coin.String()]
+	dcr.fundingMtx.RUnlock()
+	var addr string
+	if found {
+		// Try to get it from gettxout
+		addr = fCoin.addr
+	} else {
+		// Check if we can get the address from gettxout.
+		txOut, err := dcr.node.GetTxOut(&output.txHash, output.vout, true)
+		if err == nil && txOut != nil {
+			addrs := txOut.ScriptPubKey.Addresses
+			if len(addrs) != 1 {
+				return nil, nil, fmt.Errorf("multi-sig not supported")
 			}
-			pubkeys = append(pubkeys, pub.SerializeCompressed())
-			sigs = append(sigs, signature.Serialize())
+			addr = addrs[0]
+			found = true
 		}
 	}
-	if len(pubkeys) == 0 {
-		return nil, nil, fmt.Errorf("no valid keys found.")
+	// Could also try the gettransaction endpoint, which is supposed to return
+	// information about wallet transactions, but which (I think?) doesn't list
+	// ssgen outputs.
```

Comment:
looks very wrong

but listtransactions is basically wrong about everything, because it tries to list sent and received outputs, not transactions.  I recommend avoiding it.

Reply:
This output was from `gettransaction` though.  Does it suffer from similar issues?